back_ends/r/exec.R

```diff
@@ -28,24 +28,46 @@ if (file.exists(check_code_file_path)) {
   }
 }
 
-exec_jpg <- function(code) {
+exec_jpg <- function(code, i=0) {
   library(ggplot2)
 
   pdf(NULL) # Prevents Rplots.pdf from being created.
 
   eval(parse(text=code))
-  ggsave("/sandbox/image_output", dpi = 150, device = "jpeg")
+  if (i == 0) {
+    ggsave("/sandbox/image_output", dpi = 150, device = "jpeg")
+  }
+  else {
+    ggsave(paste("/sandbox/test_image_output_", i, sep=""), dpi = 150, device = "jpeg")
+  }
 }
 
 code <- readChar(code_file_path, file.info(code_file_path)$size)
 
-if (file.exists(test_code_file_path)) {
-  test_code <- readChar(test_code_file_path, file.info(test_code_file_path)$size)
-  code <- paste(code, test_code, sep="\n\n")
-}
-
 if (output_type == "txt") {
   suppressMessages(suppressWarnings(suppressPackageStartupMessages(eval(parse(text=code)))))
 } else {
   suppressMessages(suppressWarnings(suppressPackageStartupMessages(exec_jpg(code))))
 }
+
+if (dir.exists(tests_dir_path)) {
+    setwd(file.path(tests_dir_path, "outputs"))
+    outputs_dir <- getwd()
+    tests <- list.files(path=tests_dir_path, pattern="test*", full.names=TRUE, recursive=FALSE)
+    for (i in seq_along(tests)) {
+        setwd(file.path(outputs_dir, paste("test_", i, sep="")))
+        test_code <- readChar(tests[i], file.info(tests[i])$size)
+
+        # need to find a way to suppress printing to stdout while saving test code output so test outputs don't show up in submission output
+        # Need to allow for multiline output!!!! FIXME
+        if (output_type == "txt") {
+          filename <- "text_output"
+          out <- invisible(suppressMessages(suppressWarnings(suppressPackageStartupMessages(eval(parse(text=test_code))))))
```

Comment:
So I found a pretty big bug in my code here, previously we'd been printing the ```eval(parse(text=code))``` line that runs the actual code, and I assumed I could just assign the code output to a variable called ```out.``` unfortunately, this only catches the last line of output. I'm not as familiar with R so I was hoping there's some way to run the code and then save it to a variable which can be written to a file? 

Some alternatives that come to mind should that prove impossible would be not differentiating R exercise outputs by test (ie just printing the output of the solution + all tests to stdout), or going into the back end and reconfiguring things so that it hits the back end once for each test. Neither of those sound particularly optimal to me so I was hoping you would know of a better way to do things in R! Thanks!

Reply:
We may need to change the logic to be more like what we are doing in `exec.py` (python back end). We are not reading the code into a variable and then executing it. We are leaving the code in a file and using the `python` command to execute it. We should be able to do something different using the `Rscript` command. The only gotcha is that with the current code in `exec.r`, we are using the "suppress" functions to remove unnecessary output. Instead of the "suppress" functions, we could use an approach like this: https://stackoverflow.com/questions/16194212/how-to-suppress-warnings-globally-in-an-r-script.

Do you think this approach would work?